rclcpp/include/rclcpp/subscription_options.hpp

```diff
@@ -75,6 +75,70 @@ struct SubscriptionOptionsBase
   TopicStatisticsOptions topic_stats_options;
 
   QosOverridingOptions qos_overriding_options;
+
+  /// Enable use_default_callbacks setting.
+  SubscriptionOptionsBase &
+  enable_use_default_callbacks()
+  {
+    use_default_callbacks = true;
+    return *this;
+  }
+
+  /// Enable use_default_callbacks setting.
+  SubscriptionOptionsBase &
+  disable_use_default_callbacks()
+  {
+    use_default_callbacks = false;
+    return *this;
+  }
+
+  /// Enable ignore_local_publications setting.
+  SubscriptionOptionsBase &
+  enable_ignore_local_publications()
+  {
+    ignore_local_publications = true;
+    return *this;
+  }
+
+  /// Disable ignore_local_publications setting.
+  SubscriptionOptionsBase &
+  disable_ignore_local_publications()
+  {
+    ignore_local_publications = false;
+    return *this;
+  }
+
+  /// Set callback group, parameter idiom style.
+  SubscriptionOptionsBase &
+  set_callback_group(rclcpp::CallbackGroup::SharedPtr group)
```

Comment:
Note to other reviewers: we need a `set_` prefix, because we already have public members with the same name.

Should we deprecate the public members so this struct can be converted to a class in the future?
Another alternative is to apply method chaining in a builder class:

```cpp
SubscriptionOption::Builder().callback_group(group).enable_intra_process_comm().build();
```

I don't have a strong preference

Reply:
I think that given the existing member access, the builder option is best.

---

If we would prefer to change it, we could have a two tick-tock cycle to phase it out:

- ROS G:
  - new class `rclcpp::future::SubscriptionOptionsBase` with named parameter idiom properties
  - deprecate use of `rclcpp::SubscriptionOptionBase`
- ROS H:
  - move `rclcpp::future::SubscriptionOptionBase` to `rclcpp::SubscriptionOptionBase`, removing old version
  - deprecate `rclcpp::future::SubscriptionOptionBase`
- ROS I:
  - remove `rclcpp::future::SubscriptionOptionBase`

The benefit would be that we have consistency with other option-like structures in rclcpp, but the downside is the churn for developers (though we could hide it for many developers which won't touch subscription options).

Reply:
@ivanpauno what should we do here? builder I guess? Should we aim for the deprecation cycle?

@y-okumura-isp that will require moving these new functions out of this class and into a builder class instead.

Reply:
Sorry, I should have discuss API before update PR.

`SubscriptionOptionsBase` has many parameters which are `class` or `struct`, for example, `SubscriptionEventCallbacks event_callbacks` or `TopicStatisticsOptions topic_stats_options` and so on.
So I think Builder could set boolean options but we would set other parameters manually like below, right?

```
auto sub_option = SubscriptionOption::Builder().callback_group(group).use_intra_process_comm(true).build();

// we can set complex options directly
sub_option.event_callbacks.deadline_callback = [](auto) {};

// Could we overwrite parameters which Builder set? or hide them?
sub_option.use_default_callbacks = false;

// we no longer need setter functions
// sub_option.set_callback_group(group);   // no longer need
```

If above code is right, I think even if we would introduce Builder, existing API of SubscriptionOptionsBase would not change.
Would we need two tick-tock cycle in such a case?

If I misunderstand, please tell me.

Thank you.

Reply:
> So I think Builder could set boolean options but we would set other parameters manually like below, right?

I would make everything available in the builder methods, e.g.:

```cpp
auto sub_option = SubscriptionOption::Builder().callback_group(group).use_intra_process_comm(true).deadline_callback([](auto) {}).build();
```

> existing API of SubscriptionOptionsBase would not change.

That's fine, in the future we could deprecate direct access, but I think that having both options is ok.

> Would we need two tick-tock cycle in such a case?

That's needed if direct access to the fields is deprecated, which I would avoid at the moment.

Reply:
@ivanpauno Thank you for reply! I I'm sorry I missed it.

> That's fine, in the future we could deprecate direct access, but I think that having both options is ok.
> > Would we need two tick-tock cycle in such a case?
> That's needed if direct access to the fields is deprecated, which I would avoid at the moment.

I agree. I would upadte PR.